base/math.jl

```diff
@@ -94,7 +94,7 @@ julia> evalpoly(2, (1, 2, 3))
 function evalpoly(x, p::Tuple)
     if @generated
         N = length(p.parameters::Core.SimpleVector)
-        ex = :(p[end])
+        ex = :(oftype(one(x) * p[end], p[end]))
```

Comment:
I worry a bit about this causing performance regressions (especially in cases where multiplication is not cheap), but I also get that it's likely important. 

Ideally we'd be able to do something in the type domain here rather than an actual multiplication, but oh well i guess.

Reply:
Hopefully the compiler can optimize out `one(x) * p[end]` in most cases, and just compute the type, since the result is not used.

Reply:
Only if the multiplication is simple enough and free of side effects like memory allocation? If the scalar is eg a BigInt it may not be able to optimize this away?